docs/source/notes/cuda.rst

```diff
@@ -513,3 +513,189 @@ by GIL of Python interpreter.
 
 If you use :class:`~torch.nn.parallel.DistributedDataParallel`, you could use
 `torch.distributed.launch` utility to launch your program, see :ref:`distributed-launch`.
+
+.. _cuda-graph-semantics:
+
+CUDA Graphs
+-----------
+
+A graph is a frozen record of CUDA work (mostly kernels and their arguments).
+For general principles and details on the underlying CUDA API, see
+`Getting Started with CUDA Graphs`_ and the
+`Graphs section`_ of the CUDA C Programming Guide.
+
+The Pytorch integration creates graphs using `stream capture`_, which puts a CUDA stream
+in *capture mode*. CUDA work issued to a capturing stream doesn't actually run on
+the GPU. Instead, the work is recorded in a graph.
+
+After capture, the graph can be *launched* to run the GPU work as many times as needed.
+Each replay runs the same frozen ops on the same memory addresses.
+By filling input memory with new data (e.g., from a new batch) before each replay,
+you can rerun the same work on new data.
+
+Replay sacrifices the dynamic flexibility of typical eager execution in exchange for
+**greatly reduced CPU overhead**.
+
+Eager execution runs deep, often CPU-intensive Python, C++, and CUDA library call chains
+to set up arguments, dispatch them to kernels, and launch kernels one at a time.
+
+By contrast, a graph's arguments and kernels are fixed, so a graph replay elides the
+eager CPU plumbing (under the hood, a replay submits the entire graph's work to the GPU with
+a single call to `cudaGraphLaunch`_).  Kernels in a replay also execute slightly faster
+on the GPU, but eliding CPU overhead is the main benefit.
+
+You should try CUDA graphs if all or part of your network is graph-safe (usually this means
+static shapes and control flow, but see the other :ref:`constraints<capture-constraints>`)
+and you suspect its runtime is at least somewhat CPU-limited.
+
+.. _Getting Started with CUDA Graphs:
+    https://developer.nvidia.com/blog/cuda-graphs/
+.. _Graphs section:
+    https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-graphs
+.. _stream capture:
+    https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#creating-a-graph-using-stream-capture
+.. _cudaGraphLaunch:
+    https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1g1accfe1da0c605a577c22d9751a09597
+
+Pytorch API
+^^^^^^^^^^^
+
+.. warning::
+    The API presented here is a prototype and may change in future releases.
+
+Pytorch exposes graphs as a primitive :class:`torch.cuda.CUDAGraph` class,
```

Comment:
It might be nice if the previous section included a basic example showing the ideal capture -> replay pattern. Then this section could start with the recommended APIs instead of giving the reader whiplash by first introducing APIs and then explaining they're not recommended for use.

The next section, "constraints", also suffers from a lack of this example where it immediately launches into a parenthetical. An example upfront may help orient the reader and the text.

Reply:
I included an orienting example in this section, lmk what you think of the new structure.